### PR TITLE
DAG DependencyConnection draft

### PR DESCRIPTION
**Pull request by dave2s**
**MOVED HERE - > **https://github.com/MovingBlocks/Terasology/pull/3732****

Contains draft for **not yet created issue**. Drafting dependency connections for DAG
[### **Graph progress marked here, use this to navigate code too**](https://sketchboard.me/tBB7l0mTbOcg#/)

- [x] Figure EdgeConnection storing and push this forward.
- [x] Remove current mechanics for dependency storing and loading.
- [x] ~~Remove direct FBO passing. Only causes trouble.~~ still use it for reusable nodes' outputs (except bufferpairs)
- [x] Discuss naming of this feature and classes. Since the proposed naming of sockets/plugs makes more sense than connections because connections would live outside the scope of nodes and nodes would be attributed to these connections, which is not the case.
**Discuss progress**

### CODE REVIEW

**Comment by dave2s** [comment]
Do not sift trough the code until I push another update. This was first drafting and is not going to be pushed forward in this direction.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/rendering/dag/gsoc/NewAbstractNode.java [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        this.inputConnections.forEach((name, connection) -> inputConnectionNameList.add(name));
+        return inputConnectionNameList;
+    }
+
+    private List<String> getOutputConnections() {
+        List<String> outputConnectionNameList = new ArrayList<>();
+        this.inputConnections.forEach((name, connection) -> outputConnectionNameList.add(name));
+        return outputConnectionNameList;
+    }
+
```
Comment (dave2s) [inline engine/src/main/java/org/terasology/rendering/dag/gsoc/NewAbstractNode.java]: Changes from here UP

File: engine/src/main/java/org/terasology/rendering/dag/gsoc/NewAbstractNode.java [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+        this.nodeUri = new SimpleUri(providingModule.toString() + ":" + nodeId);
+        //TODO Check for empty list of either in or out
+    }
+
+    /**
+     * Each node must implement this method to be called when the node is fully connected.
+     * This method should call other private node methods to use dependencies.
+     */
+    public abstract void setDependencies(Context context);
```
Comment (dave2s) [inline engine/src/main/java/org/terasology/rendering/dag/gsoc/NewAbstractNode.java]: A bypass for using up dependencies after nodes are 1.created,2.dependencies connected

File: engine/src/main/java/org/terasology/rendering/dag/nodes/FinalPostProcessingNode.java [hunk 3]
```diff
@@ -114,7 +117,7 @@ public FinalPostProcessingNode(String nodeUri, Context context) {
         lastUpdatedGBuffer = displayResolutionDependentFBOs.getGBufferPair().getLastUpdatedFbo();
 
         int texId = 0;
-        addDesiredStateChange(new SetInputTextureFromFbo(texId++, TONE_MAPPING_FBO_URI, ColorTexture, displayResolutionDependentFBOs, POST_MATERIAL_URN, "texScene"));
+        addDesiredStateChange(new SetInputTextureFromFbo(texId++, this.getInputFboData(1), ColorTexture, displayResolutionDependentFBOs, POST_MATERIAL_URN, "texScene"));
```
Comment (dave2s) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/FinalPostProcessingNode.java]: example of new connections

File: engine/src/main/java/org/terasology/rendering/dag/nodes/FinalPostProcessingNode.java [hunk 4]
```diff
@@ -89,7 +89,10 @@
 
     public FinalPostProcessingNode(String nodeUri, Context context) {
         super(nodeUri, context);
+    }
 
+    @Override
+    public void setDependencies(Context context) {
```
Comment (dave2s) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/FinalPostProcessingNode.java]: managing everything after DependencyConnections are set

File: engine/src/main/java/org/terasology/rendering/dag/nodes/ToneMappingNode.java [hunk 5]
```diff
@@ -65,8 +65,13 @@ public ToneMappingNode(String nodeUri, Context context) {
 
         screenGrabber = context.get(ScreenGrabber.class);
 
-        DisplayResolutionDependentFBOs displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFBOs.class);
-        FBO toneMappingFbo = requiresFBO(new FBOConfig(TONE_MAPPING_FBO_URI, FULL_SCALE, FBO.Type.HDR), displayResolutionDependentFBOs);
+        DisplayResolutionDependentFBOs displayResolutionDependentFboManager = context.get(DisplayResolutionDependentFBOs.class);
+        FBO toneMappingFbo = requiresFbo(new FBOConfig(TONE_MAPPING_FBO_URI, FULL_SCALE, FBO.Type.HDR), displayResolutionDependentFboManager);
+
+        addOutputFboConnection(1,toneMappingFbo);
```
Comment (dave2s) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/ToneMappingNode.java]: add output connection inside this node
setDependencies not yet used, should be inside TODO

File: engine/src/main/java/org/terasology/rendering/dag/nodes/ToneMappingNode.java [hunk 6]
```diff
@@ -65,8 +65,13 @@ public ToneMappingNode(String nodeUri, Context context) {
 
         screenGrabber = context.get(ScreenGrabber.class);
 
-        DisplayResolutionDependentFBOs displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFBOs.class);
-        FBO toneMappingFbo = requiresFBO(new FBOConfig(TONE_MAPPING_FBO_URI, FULL_SCALE, FBO.Type.HDR), displayResolutionDependentFBOs);
+        DisplayResolutionDependentFBOs displayResolutionDependentFboManager = context.get(DisplayResolutionDependentFBOs.class);
```
Comment (dave2s) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/ToneMappingNode.java]: renamed var

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/rendering/world/WorldRendererImpl.java [hunk 7]
```diff
@@ -515,7 +515,8 @@ private void addFinalPostProcessingNodes(RenderGraph renderGraph) {
         Node initialPostProcessingNode = renderGraph.findNode("engine:initialPostProcessingNode");
         Node updateExposureNode = renderGraph.findNode("engine:updateExposureNode");
 
-        Node toneMappingNode = new ToneMappingNode("toneMappingNode", context);
+        ToneMappingNode toneMappingNode = new ToneMappingNode("toneMappingNode", context);
```
Comment (dave2s) [inline engine/src/main/java/org/terasology/rendering/world/WorldRendererImpl.java]: does it make a perf difference if we create as interface type only and cast later to get to new methods ?

File: engine/src/main/java/org/terasology/rendering/world/WorldRendererImpl.java [hunk 8]
```diff
@@ -533,7 +535,20 @@ private void addFinalPostProcessingNodes(RenderGraph renderGraph) {
...
+        FinalPostProcessingNode finalPostProcessingNode = new FinalPostProcessingNode("finalPostProcessingNode", context/*finalIn1*/);
+        /**As the last attribute - getting output dependencyUri/wholeconnection(can't see why now, but possible) from nodes - how? -
+         *                                                       either type -
+         *                                                          getOutputFBOConnection{ByNameContains}(substr/prefix/suffix..)/
+         *                                                       or getOutputFBOConnection{ByPriority}(#)-fetch output FBO
+         *                                                         - requires List instead of Map, OR adding priority attribute or...
+         *                                                       or getOutputFBOConnection{ByName}(name)*/
+        finalPostProcessingNode.connectFbo(1, toneMappingNode.getOutputFboConnection(1)); //addInputFboConnection(1, toneMappingNode.getOutputFBOUri(1));
+        finalPostProcessingNode.setDependencies(context);
```
Comment (dave2s) [inline engine/src/main/java/org/terasology/rendering/world/WorldRendererImpl.java]: Connect tone mapping output to final post process fbo input 1
Call rest of finalPostProcessing setup code

**Comment by dave2s** [comment]
Ok, i added a new update.  If you look at files changes, I added some comments to mark important changes.
This is runnable, although you can't yet test ingame FBO retrieval,a new command must be provided for retrieving FBOs in game from nodes which use new system

**Comment by Cervator** [comment]
This pull request **introduces 1 alert** when merging a3744147e8ae1ae804fe56dda54183b11a8e1e5a into dbc32f0da97e46c447ae2293a00e5babaf92c85b - [view on LGTM.com](https://lgtm.com/projects/g/MovingBlocks/Terasology/rev/pr-d0ded5303862695c4adcb5f24c8a3203c1459ca6)

**new alerts:**

* 1 for Missing space in string literal

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java [hunk 9]
```diff
@@ -166,4 +170,27 @@ public void dispose() {
         }
         nodeMap.clear();
     }
+
+    /*private void matchDependencyConnections(Node fromNode, Node toNode){
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java]: `/* private`
`) {`

File: engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java [hunk 10]
```diff
@@ -166,4 +170,27 @@ public void dispose() {
...
+        /**TODO REMOVE THIS IF WE MAKE THIS CHANGE FOR EVERY NODE
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java]: `/** TODO`

File: engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java [hunk 11]
```diff
@@ -166,4 +170,27 @@ public void dispose() {
...
+         *!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!*//*
+        if((fromNode instanceof NewAbstractNode)&&(toNode instanceof NewAbstractNode)){
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java]: `) && (`
`) {`

File: engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java [hunk 12]
```diff
@@ -166,4 +170,27 @@ public void dispose() {
...
+            NewAbstractNode fromN=(NewAbstractNode) fromNode;
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java]: `fromN = (NewAbstractNode`

File: engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java [hunk 13]
```diff
@@ -166,4 +170,27 @@ public void dispose() {
...
+            NewAbstractNode toN=(NewAbstractNode) toNode;
+
+            List<Name> fromOutputs = fromN.getOutputConnections();
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java]: Please, use toNode and fromNode. We are not working in the 80s anymore, where we have 8 length variable name limits.

File: engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java [hunk 14]
```diff
@@ -166,4 +170,27 @@ public void dispose() {
...
+
+            if(fromOutputs.size() != toInputs.size()){
+
+            }
+
+            for (Name n: fromOutputs
+                 ) {
+                
+            }
+        } else{
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java]: `else {`

File: engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java [hunk 15]
```diff
@@ -166,4 +170,27 @@ public void dispose() {
...
+        /**TODO REMOVE THIS IF WE MAKE THIS CHANGE FOR EVERY NODE
+         *!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!*//*
+        if((fromNode instanceof NewAbstractNode)&&(toNode instanceof NewAbstractNode)){
+            NewAbstractNode fromN=(NewAbstractNode) fromNode;
+            NewAbstractNode toN=(NewAbstractNode) toNode;
+
+            List<Name> fromOutputs = fromN.getOutputConnections();
+            List<Name> toInputs = toN.getInputConnections();
+
+            if(fromOutputs.size() != toInputs.size()){
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java]: `) {`

File: engine/src/main/java/org/terasology/rendering/dag/gsoc/DependencyConnection.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.terasology.rendering.dag.gsoc;
+
+
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/gsoc/DependencyConnection.java]: Extra newline

File: engine/src/main/java/org/terasology/rendering/dag/gsoc/DependencyConnection.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    private String connectionName;
+    private Type connectionType;
+
+    DependencyConnection(String name, Type type) {
+        this.connectionName = name;
+        this.connectionType = type;
+    }
+
+    public enum Type { /*Might differentiate by name only*/
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/gsoc/DependencyConnection.java]: Why are you using the multiline comment here? It looks ugly.

Ideally, use
`public enum Type { // Might differentiate by name only`
However, if you insist on multiline comments, do
`public enum Type { /* Might differentiate by name only */`

File: engine/src/main/java/org/terasology/rendering/dag/gsoc/NewAbstractNode.java [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+
+    private DependencyConnection getInputConnection(String name) {
+        return inputConnections.get(name);
+    }
+
+    private DependencyConnection getOutputConnection(String name) {
+        DependencyConnection connection = outputConnections.get(name);
+        if (connection == null) {
+            String errorMessage = String.format("Getting output connection named %s returned null. No such output connection" +
+                    "in %s", name, this.toString());
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/gsoc/NewAbstractNode.java]: Split this line on the period, and start the next one from "No such" to make the lines more self-contained and readable.

File: engine/src/main/java/org/terasology/rendering/dag/gsoc/NewAbstractNode.java [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+        return enabled;
+    }
+
+    @Override
+    public void setEnabled(boolean enabled) {
+        this.enabled = enabled;
+    }
+
+    @Override
+    public void handleCommand(String command, String... arguments) {
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/gsoc/NewAbstractNode.java]: `String ... arguments`

File: engine/src/main/java/org/terasology/rendering/world/WorldRendererImpl.java [hunk 20]
```diff
@@ -533,7 +535,20 @@ private void addFinalPostProcessingNodes(RenderGraph renderGraph) {
         LateBlurNode secondLateBlurNode = new LateBlurNode("secondLateBlurNode", context, firstLateBlurFbo, secondLateBlurFbo);
         renderGraph.addNode(secondLateBlurNode);
 
-        Node finalPostProcessingNode = new FinalPostProcessingNode("finalPostProcessingNode", context);
+        // DependencyConnection finalIn1 = DependencyConnection.createFBOConnection(1, DependencyConnection.Type.INPUT,toneMappingNode.getOutputFBOUri(1));
+        // FboConnection finalIn2 = DependencyConnection.createFBOConnection(2, DependencyConnection.Type.INPUT,secondLateBlurNode.getOutputFBOUri(2));
+        FinalPostProcessingNode finalPostProcessingNode = new FinalPostProcessingNode("finalPostProcessingNode", context/*finalIn1*/);
+        /**As the last attribute - getting output dependencyUri/wholeconnection(can't see why now, but possible) from nodes - how? -
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/world/WorldRendererImpl.java]: `/** As the`

File: engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java (hunk 13 above)
Comment (dave2s) [reply engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java]: This code is probably not going to be used anytime soon, but...help me come up with names then for future reference
`NewAbstractNode fromN=(NewAbstractNode) fromNode;  `
both are Node. Or you could just cast without saving it to a variable

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/rendering/dag/nodes/FinalPostProcessingNode.java [hunk 21]
```diff
@@ -116,10 +113,16 @@ public void setDependencies(Context context) {
...
+
+    @Override
+    public void setDependencies(Context context) {
+        DisplayResolutionDependentFBOs displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFBOs.class);
+
         int texId = 0;
         addDesiredStateChange(new SetInputTextureFromFbo(texId++, this.getInputFboData(1), ColorTexture, displayResolutionDependentFBOs, POST_MATERIAL_URN, "texScene"));
         addDesiredStateChange(new SetInputTextureFromFbo(texId++, lastUpdatedGBuffer, DepthStencilTexture, displayResolutionDependentFBOs, POST_MATERIAL_URN, "texDepth"));
-        setBlurTexture = new SetInputTextureFromFbo(texId++, SECOND_LATE_BLUR_FBO_URI, ColorTexture, displayResolutionDependentFBOs, POST_MATERIAL_URN, "texBlur");
+        setBlurTexture = new SetInputTextureFromFbo(texId++, this.getInputFboData(2), ColorTexture, displayResolutionDependentFBOs, POST_MATERIAL_URN, "texBlur");
```
Comment (dave2s) [inline engine/src/main/java/org/terasology/rendering/dag/nodes/FinalPostProcessingNode.java]: forgot to mention tone mapping input in the commit message

**Comment by dave2s** [comment]
I should store DependencyConnection connectedToConnection instead of connectedNode as DependencyConnection's attribute. Since I need to get to both connected node and it's connection...since connection now has it's parent - I can query the connection for it's parent

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

File: engine/src/main/java/org/terasology/rendering/dag/Node.java [hunk 22]
```diff
@@ -17,8 +17,13 @@
 
 //TODO: consider removing the word "Node" from the name of all Node implementations now that they are in the dag.nodes package.
 
-import java.util.Set;
+import org.terasology.context.Context;
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/Node.java]: Why these changes?

File: engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java [hunk 23]
```diff
@@ -25,66 +25,79 @@
...
-    private MutableGraph<Node> graph;
+    private Map<SimpleUri, NewNode> nodeMap;
+    private MutableGraph<NewNode> graph;
+    private Context context;
 
-    public RenderGraph() {
+    public RenderGraph(Context context) {
         nodeMap = Maps.newHashMap();
         graph = GraphBuilder.directed().build();
+        this.context = context;
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java]: Can you please tell me where this variable is used?

File: engine/src/main/java/org/terasology/rendering/dag/gsoc/PropertyChangeEvent.java [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.rendering.dag.gsoc;
+
+import org.terasology.module.sandbox.API;
+
+@API
+public class PropertyChangeEvent extends java.beans.PropertyChangeEvent {
```
Comment (vampcat) [inline engine/src/main/java/org/terasology/rendering/dag/gsoc/PropertyChangeEvent.java]: I am sorry, but what's the point of this class?

File: engine/src/main/resources/assets/shaders/tint_frag.glsl [hunk 25]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2012 Benjamin Glatzel <benjamin.glatzel@me.com>
```
Comment (vampcat) [inline engine/src/main/resources/assets/shaders/tint_frag.glsl]: What's this going to be used for?

File: engine/src/main/resources/assets/shaders/tint_vert.glsl [hunk 26]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright 2012 Benjamin Glatzel <benjamin.glatzel@me.com>
```
Comment (vampcat) [inline engine/src/main/resources/assets/shaders/tint_vert.glsl]: What's this going to be used for?

Comment (dave2s) [reply engine/src/main/resources/assets/shaders/tint_vert.glsl]: mistakenly copypasted copyright, shader should've been moved to module repo, will remove file

File: engine/src/main/resources/assets/shaders/tint_frag.glsl (hunk 25 above)
Comment (dave2s) [reply engine/src/main/resources/assets/shaders/tint_frag.glsl]: mistakenly copypasted copyright, shader should've been moved to module repo, will remove file

File: engine/src/main/java/org/terasology/rendering/dag/gsoc/PropertyChangeEvent.java (hunk 24 above)
Comment (dave2s) [reply engine/src/main/java/org/terasology/rendering/dag/gsoc/PropertyChangeEvent.java]: This class probably isn't needed, but I still haven't tested the events for module based class. There's this problem when using 3rd party library during runtime from within a module...forgot I could maybe whitelist it? I'll take a look

Comment (vampcat) [reply engine/src/main/java/org/terasology/rendering/dag/gsoc/PropertyChangeEvent.java]: Yes, by default you can't use 3rd party libraries from modules. However, you can whitelist them explicitly. Way cleaner.

File: engine/src/main/java/org/terasology/rendering/dag/gsoc/RenderDagApi.java [hunk 27]
```diff
@@ -0,0 +1,233 @@
...
+ * Singleton class which is used by modules trough {@link RenderDagApiInterface} obtained from {@link Context}.
+ * </p>
+ * <p>
+ * Implements basic operations over DAG and dependency passing.
+ * </p>
+ */
+public final class RenderDagApi implements RenderDagApiInterface {
+    private static final Logger logger = LoggerFactory.getLogger(RenderDagApi.class);
+
+    private static RenderDagApi singleInstance;
```
[the remark this answers is not in the document]
Comment (dave2s) [reply engine/src/main/java/org/terasology/rendering/dag/gsoc/RenderDagApi.java]: I need to interweave dependency connections with render graph connections. I got no response back then so I went with another class for this logic...I'll need to think this over

File: engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java (hunk 23 above)
Comment (dave2s) [reply engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java]: I still haven't decided where to store context to use in nodes...we'll see what I wind up with after figuring out how to merge some classes. At this point though it is probably redundant (Node.setDependencies() - node has the context)

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by Cervator** [comment]
This pull request **introduces 1 alert** when merging 18236a28568ad4805bed017049fcf24cfa04465b into 07cda61532a656c9fec148e1eee565640f0bcb9a - [view on LGTM.com](https://lgtm.com/projects/g/MovingBlocks/Terasology/rev/pr-dae69222a1172a3cfdc07b4dac8a4e1865f6c00e)

**new alerts:**

* 1 for Equals on incomparable types

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by Cervator** [comment]
This pull request **introduces 1 alert** when merging a8f74b9a258a7adc07feffc208dd03e365e0c6ff into 07cda61532a656c9fec148e1eee565640f0bcb9a - [view on LGTM.com](https://lgtm.com/projects/g/MovingBlocks/Terasology/rev/pr-a26d892d2db5caba57d9a442dc9a7ff459af0f99)

**new alerts:**

* 1 for Equals on incomparable types

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by dave2s** [comment]
image

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by dave2s** [comment]
last update - change naming of module to module classes name + module name
image

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by dave2s** [comment]
MOVED HERE - > #3732
